<h1> My Coursera Capstone Project <h2>

<h3> Predicting the Severity of a Potential Accident <h3>

This notebook will be mainly used for the Coursera capstone project

In [54]:
import pandas as pd

In [55]:
import numpy as np

In [56]:
print("Hello Capstone Project Course!")

Hello Capstone Project Course!


<b> About the Data <b>

Business Understanding:
The initial phase is to understand the project's objective from the business or application perspective. Then, you need to translate this knowledge into a machine learning problem with a preliminary plan to achieve the objectives.

Data understanding:
In this phase, you need to collect or extract the dataset from various sources such as csv file or SQL database. Then, you need to determine the attributes (columns) that you will use to train your machine learning model. Also, you will assess the condition of chosen attributes by looking for trends, certain patterns, skewed information, correlations, and so on.

Data Preparation:
The data preparation includes all the required activities to construct the final dataset which will be fed into the modeling tools. Data preparation can be performed multiple times and it includes balancing the labeled data, transformation, filling missing data, and cleaning the dataset.

Modeling:
In this phase, various algorithms and methods can be selected and applied to build the model including supervised machine learning techniques. You can select SVM, XGBoost, decision tree, or any other techniques. You can select a single or multiple machine learning models for the same data mining problem. At this phase, stepping back to the data preparation phase is often required.

Evaluation:
Before proceeding to the deployment stage, the model needs to be evaluated thoroughly to ensure that the business or the applications' objectives are achieved. Certain metrics can be used for the model evaluation such as accuracy, recall, F1-score, precision, and others.

Deployment:
The deployment phase requirements vary from project to project. It can be as simple as creating a report, developing interactive visualization, or making the machine learning model available in the production environment. In this environment, the customers or end-users can utilize the model in different ways such as API, website, or so on.

This data set is about accident (car collisions) severity. This data includes all types of collisions. Collisions will display at the intersection or mid-block of a segment. The data dates weekly from 2004 to present. The data has been collected from the Seattle Department of Transportation.

<b> Load Data from CSV file <b>

In [60]:
# The code was removed by Watson Studio for sharing.

SEVERITYCODE           X          Y  OBJECTID  INCKEY  COLDETKEY REPORTNO  \
0             2 -122.323148  47.703140         1    1307       1307  3502005   
1             1 -122.347294  47.647172         2   52200      52200  2607959   
2             1 -122.334540  47.607871         3   26700      26700  1482393   
3             1 -122.334803  47.604803         4    1144       1144  3503937   
4             2 -122.306426  47.545739         5   17700      17700  1807429   

    STATUS      ADDRTYPE   INTKEY  ... ROADCOND                LIGHTCOND  \
0  Matched  Intersection  37475.0  ...      Wet                 Daylight   
1  Matched         Block      NaN  ...      Wet  Dark - Street Lights On   
2  Matched         Block      NaN  ...      Dry                 Daylight   
3  Matched         Block      NaN  ...      Dry                 Daylight   
4  Matched  Intersection  34387.0  ...      Wet                 Daylight   

  PEDROWNOTGRNT  SDOTCOLNUM SPEEDING ST_COLCODE  \
0           NaN         NaN      NaN         10   
1           NaN   6354039.0      NaN         11   
2           NaN   4323031.0      NaN         32   
3           NaN         NaN      NaN         23   
4           NaN   4028032.0      NaN         10   

                                          ST_COLDESC  SEGLANEKEY  \
0                                  Entering at angle           0   
1  From same direction - both going straight - bo...           0   
2                             One parked--one moving           0   
3                   From same direction - all others           0   
4                                  Entering at angle           0   

   CROSSWALKKEY  HITPARKEDCAR  
0             0             N  
1             0             N  
2             0             N  
3             0             N  
4             0             N  

[5 rows x 38 columns]

In [64]:
df_data_1.shape

(194673, 38)

This data set contains 194673 rows. There are 37 attributes, some numeric and some categorical. Some attributes have missing data. This is not a balanced labeled dataset.

In [65]:
df_data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194673 entries, 0 to 194672
Data columns (total 38 columns):
SEVERITYCODE      194673 non-null int64
X                 189339 non-null float64
Y                 189339 non-null float64
OBJECTID          194673 non-null int64
INCKEY            194673 non-null int64
COLDETKEY         194673 non-null int64
REPORTNO          194673 non-null object
STATUS            194673 non-null object
ADDRTYPE          192747 non-null object
INTKEY            65070 non-null float64
LOCATION          191996 non-null object
EXCEPTRSNCODE     84811 non-null object
EXCEPTRSNDESC     5638 non-null object
SEVERITYCODE.1    194673 non-null int64
SEVERITYDESC      194673 non-null object
COLLISIONTYPE     189769 non-null object
PERSONCOUNT       194673 non-null int64
PEDCOUNT          194673 non-null int64
PEDCYLCOUNT       194673 non-null int64
VEHCOUNT          194673 non-null int64
INCDATE           194673 non-null object
INCDTTM           194673 non-null obje

In [66]:
df_data_1.describe(include="all")

SEVERITYCODE              X              Y       OBJECTID  \
count   194673.000000  189339.000000  189339.000000  194673.000000   
unique            NaN            NaN            NaN            NaN   
top               NaN            NaN            NaN            NaN   
freq              NaN            NaN            NaN            NaN   
mean         1.298901    -122.330518      47.619543  108479.364930   
std          0.457778       0.029976       0.056157   62649.722558   
min          1.000000    -122.419091      47.495573       1.000000   
25%          1.000000    -122.348673      47.575956   54267.000000   
50%          1.000000    -122.330224      47.615369  106912.000000   
75%          2.000000    -122.311937      47.663664  162272.000000   
max          2.000000    -122.238949      47.734142  219547.000000   

               INCKEY      COLDETKEY REPORTNO   STATUS ADDRTYPE  \
count   194673.000000  194673.000000   194673   194673   192747   
unique            NaN            NaN   194670        2        3   
top               NaN            NaN  1780512  Matched    Block   
freq              NaN            NaN        2   189786   126926   
mean    141091.456350  141298.811381      NaN      NaN      NaN   
std      86634.402737   86986.542110      NaN      NaN      NaN   
min       1001.000000    1001.000000      NaN      NaN      NaN   
25%      70383.000000   70383.000000      NaN      NaN      NaN   
50%     123363.000000  123363.000000      NaN      NaN      NaN   
75%     203319.000000  203459.000000      NaN      NaN      NaN   
max     331454.000000  332954.000000      NaN      NaN      NaN   

               INTKEY  ... ROADCOND LIGHTCOND PEDROWNOTGRNT    SDOTCOLNUM  \
count    65070.000000  ...   189661    189503          4667  1.149360e+05   
unique            NaN  ...        9         9             1           NaN   
top               NaN  ...      Dry  Daylight             Y           NaN   
freq              NaN  ...   124510    116137          4667           NaN   
mean     37558.450576  ...      NaN       NaN           NaN  7.972521e+06   
std      51745.990273  ...      NaN       NaN           NaN  2.553533e+06   
min      23807.000000  ...      NaN       NaN           NaN  1.007024e+06   
25%      28667.000000  ...      NaN       NaN           NaN  6.040015e+06   
50%      29973.000000  ...      NaN       NaN           NaN  8.023022e+06   
75%      33973.000000  ...      NaN       NaN           NaN  1.015501e+07   
max     757580.000000  ...      NaN       NaN           NaN  1.307202e+07   

       SPEEDING ST_COLCODE              ST_COLDESC     SEGLANEKEY  \
count      9333     194655                  189769  194673.000000   
unique        1        115                      62            NaN   
top           Y         32  One parked--one moving            NaN   
freq       9333      27612                   44421            NaN   
mean        NaN        NaN                     NaN     269.401114   
std         NaN        NaN                     NaN    3315.776055   
min         NaN        NaN                     NaN       0.000000   
25%         NaN        NaN                     NaN       0.000000   
50%         NaN        NaN                     NaN       0.000000   
75%         NaN        NaN                     NaN       0.000000   
max         NaN        NaN                     NaN  525241.000000   

        CROSSWALKKEY  HITPARKEDCAR  
count   1.946730e+05        194673  
unique           NaN             2  
top              NaN             N  
freq             NaN        187457  
mean    9.782452e+03           NaN  
std     7.226926e+04           NaN  
min     0.000000e+00           NaN  
25%     0.000000e+00           NaN  
50%     0.000000e+00           NaN  
75%     0.000000e+00           NaN  
max     5.239700e+06           NaN  

[11 rows x 38 columns]

In [67]:
df_data_1['ROADCOND'].isna().sum()

5012

<b> Data Preparation <b>

In [68]:
df_data_1['INCDATE'].head()

0    2013/03/27 00:00:00+00
1    2006/12/20 00:00:00+00
2    2004/11/18 00:00:00+00
3    2013/03/29 00:00:00+00
4    2004/01/28 00:00:00+00
Name: INCDATE, dtype: object

In [72]:
df_data_2 = df_data_1.copy()

In [73]:
df_data_1.columns

Index(['SEVERITYCODE', 'X', 'Y', 'OBJECTID', 'INCKEY', 'COLDETKEY', 'REPORTNO',
       'STATUS', 'ADDRTYPE', 'INTKEY', 'LOCATION', 'EXCEPTRSNCODE',
       'EXCEPTRSNDESC', 'SEVERITYCODE.1', 'SEVERITYDESC', 'COLLISIONTYPE',
       'PERSONCOUNT', 'PEDCOUNT', 'PEDCYLCOUNT', 'VEHCOUNT', 'INCDATE',
       'INCDTTM', 'JUNCTIONTYPE', 'SDOT_COLCODE', 'SDOT_COLDESC',
       'INATTENTIONIND', 'UNDERINFL', 'WEATHER', 'ROADCOND', 'LIGHTCOND',
       'PEDROWNOTGRNT', 'SDOTCOLNUM', 'SPEEDING', 'ST_COLCODE', 'ST_COLDESC',
       'SEGLANEKEY', 'CROSSWALKKEY', 'HITPARKEDCAR'],
      dtype='object')

In [74]:
df_data_2.columns = ['Severity_Code', 'X', 'Y', 'Object_ID', 'Inc_Key', 'Coldet_Key', 'Report_No',
       'Status', 'Addr_Type', 'Int_Key', 'Location', 'Exceptrsn_Code',
       'Exceptrsn_Desc', 'Severity_Code.1', 'Severity_Desc', 'Collision_Type',
       'Person_Count', 'Ped_Count', 'Ped_Cyl_Count', 'Veh_Count', 'Inc_Date',
       'INC_DTTM', 'Junction_Type', 'SDOT_Col_Code', 'SDOT_Col_Desc',
       'Inattention_Ind', 'Under_Infl', 'Weather', 'Road_Cond', 'Light_Cond',
       'Ped_ROW_Not_Grnt', 'SDOT_Col_Num', 'Speeding', 'ST_Col_Code', 'ST_Col_Desc',
       'Seg_Lane_Key', 'Crosswalk_Key', 'Hit_Parked_Car']
df_data_2.head()

Severity_Code           X          Y  Object_ID  Inc_Key  Coldet_Key  \
0              2 -122.323148  47.703140          1     1307        1307   
1              1 -122.347294  47.647172          2    52200       52200   
2              1 -122.334540  47.607871          3    26700       26700   
3              1 -122.334803  47.604803          4     1144        1144   
4              2 -122.306426  47.545739          5    17700       17700   

  Report_No   Status     Addr_Type  Int_Key  ... Road_Cond  \
0   3502005  Matched  Intersection  37475.0  ...       Wet   
1   2607959  Matched         Block      NaN  ...       Wet   
2   1482393  Matched         Block      NaN  ...       Dry   
3   3503937  Matched         Block      NaN  ...       Dry   
4   1807429  Matched  Intersection  34387.0  ...       Wet   

                Light_Cond Ped_ROW_Not_Grnt  SDOT_Col_Num Speeding  \
0                 Daylight              NaN           NaN      NaN   
1  Dark - Street Lights On              NaN     6354039.0      NaN   
2                 Daylight              NaN     4323031.0      NaN   
3                 Daylight              NaN           NaN      NaN   
4                 Daylight              NaN     4028032.0      NaN   

  ST_Col_Code                                        ST_Col_Desc  \
0          10                                  Entering at angle   
1          11  From same direction - both going straight - bo...   
2          32                             One parked--one moving   
3          23                   From same direction - all others   
4          10                                  Entering at angle   

   Seg_Lane_Key  Crosswalk_Key  Hit_Parked_Car  
0             0              0               N  
1             0              0               N  
2             0              0               N  
3             0              0               N  
4             0              0               N  

[5 rows x 38 columns]